ontology/obo/tree.go

```diff
@@ -0,0 +1,160 @@
+	termMap := makeTermMap(terms)
+
+	var parentId string
+	var foundInMap bool
+	var parentTerm *Obo
+	var isAItem IsADescription
+
+	// Now we iterate through terms and set up the tree
+	for i := range terms {
+		term := &terms[i] //get a memory copy
+		for _, isAItem = range term.IsA {
+			parentId = isAItem.ParentId
+			if parentTerm, foundInMap = termMap[parentId]; foundInMap {
+				term.Parents = append(term.Parents, parentTerm)
+				parentTerm.Children = append(parentTerm.Children, term)
+			} else {
+				//log.Fatalf("Error: The term with ID \"%s\" has a parent with ID \"%s\" that "+
+				//	"is not found in the Obo file.\n", term.Id, parentId)
+			}
+		}
+	}
+	return termMap
+}
+
+// termToDot is a helper function of ToDot that converts and individual Obo struct to a
+// DOT format digraph descriptor, and recursively does so for all children of the input Obo struct.
+func termToDot(term Obo, out *fileio.EasyWriter, visitedNode map[string]bool) {
+	var err error
+	if visitedNode[term.Id] {
+		return
+	}
+	visitedNode[term.Id] = true
+	_, err = fmt.Fprintf(out, "\"%s\" [label = \"%s\"];\n", term.Id, term.Name)
+	exception.PanicOnErr(err)
+
+	// now we recursively visit children
+	for _, child := range term.Children {
+		_, err = fmt.Fprintf(out, "\"%s\" -> \"%s\";\n", term.Id, child.Id)
+		termToDot(*child, out, visitedNode)
+	}
+}
+
+// ToDot formats an input Obo slice as a tree and writes the resulting tree to a file in DOT format.
+func ToDot(outFile string, terms []Obo) {
+	var err error
+	out := fileio.EasyCreate(outFile)
+	BuildTree(terms)
+
+	//write header line for Dot
+	_, err = fmt.Fprintf(out, "digraph G{\n")
+	exception.PanicOnErr(err)
+
+	var visitedNode = make(map[string]bool)
+
+	for _, term := range terms {
+		termToDot(term, out, visitedNode)
+	}
+
+	//write tail line for Dot and close file
+	_, err = fmt.Fprintf(out, "}\n")
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+// makeTermMap creates a map to find nodes based on their ID and AltIDs
+func makeTermMap(terms []Obo) map[string]*Obo {
+	var altId string
+	termMap := make(map[string]*Obo)
+	for i := range terms {
+		term := &terms[i] // get a memory copy
+		if termMap[term.Id] != nil {
+			log.Fatalf("Term \"%v\" is duplicated in input slice of Obo.\n", term.Id)
+		}
+		termMap[term.Id] = term
+		for _, altId = range term.AltIds {
+			if termMap[altId] != nil {
+				log.Fatalf("Term \"%v\" is duplicated in input slice of Obo.\n", term.Id)
+			}
+			termMap[altId] = term
+		}
+	}
+
+	return termMap
+}
+
+// SubtreeToDot makes a DOT format tree for the subtree rooted at the input nodeId.
+// BuildTree must have been run on this tree already, as termMap is an argument.
+func SubtreeToDot(outFile string, nodeId string, termMap map[string]*Obo) {
+	var err error
+	out := fileio.EasyCreate(outFile)
+	//write header line for Dot
+	_, err = fmt.Fprintf(out, "digraph G {\n")
+	exception.PanicOnErr(err)
+
+	var visitedNode = make(map[string]bool)
```

Comment:
This is another place where we have some added complexity because of storing a tree in a list data structure.